// @antfu/ni@0.18.8/dist/index.cjs
'use strict';

const runner = require('./shared/ni.a8753743.cjs');
require('path');
require('readline');
require('events');
require('node:buffer');
require('node:path');
require('node:child_process');
require('node:process');
require('child_process');
require('fs');
require('node:url');
require('os');
require('node:os');
require('assert');
require('buffer');
require('stream');
require('util');
require('node:fs');
require('tty');



exports.cmdExists = runner.cmdExists;
exports.detect = runner.detect;
exports.exclude = runner.exclude;
exports.getCommand = runner.getCommand;
exports.getConfig = runner.getConfig;
exports.getDefaultAgent = runner.getDefaultAgent;
exports.getGlobalAgent = runner.getGlobalAgent;
exports.getVoltaPrefix = runner.getVoltaPrefix;
exports.parseNa = runner.parseNa;
exports.parseNi = runner.parseNi;
exports.parseNr = runner.parseNr;
exports.parseNu = runner.parseNu;
exports.parseNun = runner.parseNun;
exports.parseNx = runner.parseNx;
exports.remove = runner.remove;
exports.run = runner.run;
exports.runCli = runner.runCli;
