// stag-relayx@0.2.8/dist/cjs/actor.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Actor = exports.authorIdentityPrefix = void 0;
const bsv = __importStar(require("bsv-2"));
const uuid = __importStar(require("uuid"));
exports.authorIdentityPrefix = '15PciHG22SNLQJXMoSUaWVi7WSqc7hCfva';
class Actor {
    constructor({ relayone }) {
        this.relayone = relayone;
    }
    get identity() {
        return '';
    }
    publishOpReturn(opReturn) {
        return __awaiter(this, void 0, void 0, function* () {
        });
    }
    publishMessage(newMessage, sign = false) {
        return __awaiter(this, void 0, void 0, function* () {
            newMessage.nonce = newMessage.nonce || uuid.v4();
            const payloadToSign = JSON.stringify(Object.assign(newMessage.content, {
                _app: newMessage.app,
                _type: newMessage.type,
                _nonce: newMessage.nonce
            }));
            console.log("PAYLOAD TO SIGN", payloadToSign);
            let signature = '';
            const send = {
                to: 'johngalt@relayx.io',
                amount: 0.001,
                currency: 'BSV',
                opReturn: [
                    'onchain.sv',
                    newMessage.app,
                    newMessage.type,
                    payloadToSign,
                    "|",
                    exports.authorIdentityPrefix,
                    "BITCOIN_ECDSA",
                    this.identity,
                    signature,
                    '0x05' // signed index #5 "payloadToSign"
                ]
            };
            console.log("relayone.send", send);
            return this.relayone.send(send);
        });
    }
    socialMessage(newMessage) {
        const defaultContentType = 'text/markdown';
        const contentType = newMessage.contentType || 'text/markdown';
        return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
            const payloadToSign = newMessage.content;
            let address = new bsv.Address().fromString(this.identity);
            var signature = '';
            return this.publishOpReturn([
                'B',
                newMessage.app,
                payloadToSign,
                "|",
                exports.authorIdentityPrefix,
                "BITCOIN_ECDSA",
                this.identity,
                signature,
                '0x05' // signed index #5 "payloadToSign"
            ]);
            let verified = bsv.Bsm.verify(Buffer.from(payloadToSign, 'utf8'), signature, address);
            if (!verified) {
                throw new Error('SIGNATURE NOT VERIFIED');
            }
        }));
    }
}
exports.Actor = Actor;
