// markdown-it-shiki@0.7.2/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  default: () => src_default,
  resolveOptions: () => resolveOptions
});
module.exports = __toCommonJS(src_exports);

// node_modules/.pnpm/tsup@6.5.0_typescript@4.9.3/node_modules/tsup/assets/cjs_shims.js
var getImportMetaUrl = () => typeof document === "undefined" ? new URL("file:" + __filename).href : document.currentScript && document.currentScript.src || new URL("main.js", document.baseURI).href;
var importMetaUrl = /* @__PURE__ */ getImportMetaUrl();

// src/index.ts
var import_module = require("module");
var import_synckit = require("synckit");
function getThemeName(theme) {
  if (typeof theme === "string")
    return theme;
  return theme.name;
}
var RE = /{([\d,-]+)}/;
function resolveOptions(options) {
  const themes = [];
  let darkModeThemes;
  if (!options.theme) {
    themes.push("nord");
  } else if (typeof options.theme === "string") {
    themes.push(options.theme);
  } else {
    if ("dark" in options.theme || "light" in options.theme) {
      darkModeThemes = options.theme;
      themes.push(darkModeThemes.dark);
      themes.push(darkModeThemes.light);
    } else {
      themes.push(options.theme);
    }
  }
  return {
    ...options,
    themes,
    darkModeThemes: darkModeThemes ? {
      dark: getThemeName(darkModeThemes.dark),
      light: getThemeName(darkModeThemes.light)
    } : void 0
  };
}
var attrsToLines = (attrs) => {
  const result = [];
  if (!attrs.trim())
    return [];
  attrs.split(",").map((v) => v.split("-").map((v2) => parseInt(v2, 10))).forEach(([start, end]) => {
    if (start && end) {
      result.push(
        ...Array.from({ length: end - start + 1 }, (_, i) => start + i)
      );
    } else {
      result.push(start);
    }
  });
  return result.map((v) => ({
    line: v,
    classes: ["highlighted"]
  }));
};
var MarkdownItShiki = (markdownit, options = {}) => {
  const _highlighter = options.highlighter;
  const {
    langs,
    themes,
    darkModeThemes,
    highlightLines
  } = resolveOptions(options);
  let syncRun;
  if (!_highlighter) {
    const require2 = (0, import_module.createRequire)(importMetaUrl);
    syncRun = (0, import_synckit.createSyncFn)(require2.resolve("./worker"));
    syncRun("getHighlighter", { langs, themes });
  }
  const highlightCode = (code, lang, theme, lineOptions) => {
    if (_highlighter)
      return _highlighter.codeToHtml(code, { lang: lang || "text", theme, lineOptions });
    return syncRun("codeToHtml", {
      code,
      theme,
      lang: lang || "text",
      lineOptions
    });
  };
  markdownit.options.highlight = (code, lang, attrs) => {
    let lineOptions;
    if (highlightLines) {
      const match = RE.exec(attrs);
      if (match)
        lineOptions = attrsToLines(match[1]);
    }
    if (darkModeThemes) {
      const dark = highlightCode(code, lang, darkModeThemes.dark, lineOptions).replace('<pre class="shiki"', '<pre class="shiki shiki-dark"');
      const light = highlightCode(code, lang || "text", darkModeThemes.light, lineOptions).replace('<pre class="shiki"', '<pre class="shiki shiki-light"');
      return `<div class="shiki-container">${dark}${light}</div>`;
    } else {
      return highlightCode(code, lang || "text");
    }
  };
};
var src_default = MarkdownItShiki;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  resolveOptions
});
